Adding eventhandlers to classes of units in SQF

commy2 opened the pull request:
This PR adds CBA_fnc_addClassEventhandler.

Idea behind this:
Extended_xx_Eventhandlers are usefull for addon and mission makers, because they fill a specific niche: Adding eventhandlers to entire classes instead of individual units.
The SQF scripting command `addEventhandler` unfortunately only works on individual objects (which have to already exist) and the `Eventhandlers` class is often used by BI or addon makers and overwrites an eventhandler already added in, for example `CAManBase`.
Extended_xx_Eventhandlers offers a solution for this, but relies on the addon makers to not overwrite the config entries from CBA too. Eventhandlers that are added by BI have to be recreated using XEH and these classes have to be fixed manually. Sometimes eventhandlers are added in new game versions without the changelog ever mentioning it. (remember the FIA soldiers?)
This new function adds all eventhandlers in SQF. It does so by polling all entities (`entities ""`). It does not need any configs at all and works basically on any mission created object.

useage:
["CAManBase", "fired", {systemChat str _this}] call CBA_fnc_addClassEventhandler;
["All", "init", {systemChat str _this}] call CBA_fnc_addClassEventhandler;

advantages:
- no configs need
- way less maintenance needed
- automatically compatible with everything (no auto_xeh needed for incompatible addons)
- cleaner code with less overhead compared to XEH (my opinion)

disadvantages:
- polling (auto_xeh does this anyway)
- eventhandlers could be accidently removed with `removeAllEventhandlers`
- no backwards compatibility with XEH

Polling shouldn't be an issue really. `entities ""` takes 0.005896ms (on my machine) and is ~1000x times faster than `allMissionObjects ""` (although something seems of with that command). Comparing the arrays is really fast too due to `isEqualTo` being a binary comparision even with a large number of objects present. The order of the entities seems to never change for no reason, so useless iterations shouldn't happen.

notes:
Due to the nature of polling, init eventhandlers are called one frame after the object is created. Similar things happen with auto_xeh. This way it at least is predictable behaviour, although the config init eventhandler behaver slightly different.
Init eventhandlers have to be put in init box, XEH_preInit, or a XEH_postInit when the PBO is loaded after this (required addons).

jonpas commented:
Note: this is using `private` keyword so it won't work on stable before 1.54.

jokoho48 commented on addons/class_eventhandlers/XEH_postInit.sqf:
```diff
@@ -0,0 +1,51 @@
+GVAR(entities) = [];
+
+[{
+    if !(entities "" isEqualTo GVAR(entities)) then {
```
why you call twice entities and not save it in a variable? like that
```
private _entities = entities "";
if !(_entities isEqualTo GVAR(entities)) then {
    GVAR(entities) = _entities;
```

commy2 replied:
Cause 99% of frames won't have the value changed, so overall cost is lower.

jokoho48 commented on addons/class_eventhandlers/XEH_postInit.sqf:
```diff
@@ -0,0 +1,51 @@
+            false
+        } count _entities;
+    };
+}, 0, []] call CBA_fnc_addPerFrameHandler;
```
why this musst run in a PFH that get checked everyframe? this take way to much performance on my side

commy2 replied:
I didn't notice any performance cost in my tests.

Killswitch00 replied:
Don't start the PFH unless/until 'addClassEventhandler' has been used.

commy2 commented:
Using the old private syntax now so this works with 1.52.
Polling loop now only starts after the first eventhandler is added and added a 0.1 seconds sleep.
Also renamed it to `CBA_fnc_addEventHandlerToClass` because I like that name more.

jonpas commented:
We'll have to update it to 1.54 syntax anyways, because I used `local` in my cargo removal PRs.

Killswitch00 commented:
Looks better now, but wasn't the old name better? BIS have named their own event handler addition methods add<SOMETHING>EventHandler:
```
add3DENEventHandler
addEventHandler
addMissionEventHandler
addMPEeventHandler
addMusicEventHandler
addPublicVariableEventHandler
```
CBA_fnc_**addClassEventHandler** would fit in nicely there.

(Granted, if one looks at `ctrlAddEventHandler` and  `displayAddEventhandler` a case could be made for `CBA_fnc_classAddEventHandler` as well...)

commy2 commented:
Good point. I think that was what I was going for initially.
